### PR TITLE
Add EIP: Assemble assets into NFTs

### PR DESCRIPTION
**Pull request by Ungigdu**
This standard defines a token which can presents a combination of assets.

### CODE REVIEW

**Comment by alita-moore** [comment]
Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):


	 - File with name EIPS/eip-draft_assembly_nft.md is new and new files must be reviewed
	 - This PR requires review from one of [@MicahZoltu, @lightclient]
 @MicahZoltu @lightclient

**Comment by Ungigdu** [comment]
> Hi! I'm a bot, and I wanted to automerge your PR, but couldn't because of the following issue(s):
> 
> ```
>  - File with name EIPS/eip-3589.md is new and new files must be reviewed
>  - This PR requires review from one of [@MicahZoltu, @lightclient]
> ```
> 
> @MicahZoltu @lightclient

Should I wait?

**Comment by MicahZoltu** [comment]
> Should I wait?

Yeah.  I don't handle ERCs, so you'll need to wait for an editor who reviews ERCs to get a chance to take a look (it can sometimes take a while as we are short staffed).

File: EIPS/eip-3589.md [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+discussions-to: https://github.com/ethereum/EIPs/issues/3590
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-05-24
+requires: 721
+---
+
+## Simple Summary
+This standard defines a token which can presents a combination of assets.
```
Comment (lightclient) [inline EIPS/eip-3589.md:14]:
```suggestion
This standard defines a token which can represent a combination of assets.
```
please consider this update for clarity

File: EIPS/eip-3589.md [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+
+## Abstract
+The ERC-1155 multi token contract defines a way to batch transfer tokens, but those tokens must be mint by the ERC-1155 contract itself. This eip is an ERC-721 extension with ability to assemble assets such as Eth, ERC-20 tokens, ERC-721 tokens and ERC-1155 tokens into one ERC-721 token whose token id is also the assets signature. As assets get assembled into one, batch transfer or swap can be implemented very easily.
```
Comment (lightclient) [inline EIPS/eip-3589.md:17]:
```suggestion
The ERC-1155 multi-token contract defines a way to batch transfer tokens, but those tokens must be minted by the ERC-1155 contract itself. This EIP is an ERC-721 extension with ability to assemble assets such as ether, ERC-20 tokens, ERC-721 tokens and ERC-1155 tokens into one ERC-721 token whose token id is also the asset's signature. As assets get assembled into one, batch transfer or swap can be implemented very easily.
```

File: EIPS/eip-3589.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+## Motivation
+As NFT arts and collectors rapidly increases, some collectors are not satisfied with traditional trading methods. When two collectors want to swap some of their collections, currently they can list their NFTs on market and notify other peer to buy, but this is inefficient and gas consuming. Instead, some collectors turn to social media or chart group looking for a trustworthy third party to swap NFTs for them. The third party takes NFTs from both collector A and B, and transfer A's collections to B and B's to A. This is very risky.
```
Comment (lightclient) [inline EIPS/eip-3589.md:20]:
```suggestion
As NFT arts and collectors rapidly increases, some collectors are not satisfied with traditional trading methods. When two collectors want to swap some of their collections, currently they can list their NFTs on the market and notify the other party to buy, but this is inefficient and gas-intensive. Instead, some collectors turn to social media or chat group looking for a trustworthy third party to swap NFTs for them. The third party takes NFTs from both collector A and B, and transfer A's collections to B and B's to A. This is very risky.
```

File: EIPS/eip-3589.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+The safest way to do batch swap, is to transform batch swap into atomic swap, i.e. one to one swap. But first we should "assemble" those Eth, ERC-20 tokens, ERC-721 tokens and ERC-1155 tokens together, and this is the main purpose of this eip.
```
Comment (lightclient) [inline EIPS/eip-3589.md:22]:
```suggestion
The safest way to do batch swap, is to transform batch swap into atomic swap, i.e. one to one swap. But first we should "assemble" those ether, ERC-20 tokens, ERC-721 tokens and ERC-1155 tokens together, and this is the main purpose of this EIP.
```

File: EIPS/eip-3589.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+      for(uint256 k =0; k< _numbers[3]; k++){
+        IERC1155(_addresses[pointerA++]).safeTransferFrom(_msgSender(), address(this), _numbers[pointerB], _numbers[_numbers[3] + pointerB++], "");
+      }
+      tokenId = hash(nonce, _addresses, _numbers);
+      super._mint(_to, tokenId);
+      emit AssemblyAsset(_to, tokenId, nonce, _addresses, _numbers);
+      nonce ++;
+  }
+
+  function sateMint(address _to, address[] memory _addresses, uint256[] memory _numbers) payable external override returns(uint256 tokenId){
```
Comment (lightclient) [inline EIPS/eip-3589.md:147]:
```suggestion
  function safeMint(address _to, address[] memory _addresses, uint256[] memory _numbers) payable external override returns(uint256 tokenId){
```

**Review by lightclient — CHANGES_REQUESTED** [review]
Please review my comments. There are several grammatical errors in this EIP. I recommend going it over a few more times to avoid any other grammatical errors or types. Thanks!

**Comment by Ungigdu** [comment]
> Please review my comments. There are several grammatical errors in this EIP. I recommend going it over a few more times to avoid any other grammatical errors or types. Thanks!

I fixed those grammatical errors, thank you! Need to imporve my english😊

**Comment by Ungigdu** [comment]
> Please review my comments. There are several grammatical errors in this EIP. I recommend going it over a few more times to avoid any other grammatical errors or types. Thanks!

I fixed those grammatical errors, thank you! Need to imporve my english😊

File: EIPS/eip-3589.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+There are many reasons why people want to pack their NFTs together. For example, a collector want to pack a set of football players into a football team; a collector has hundreds of of NFTs with no categories to manage them; a collector wants to buy a full collection of NFTs or none of them. They all need a way a assemble those NFTs together.
+
+The reason for choosing ERC-721 standard as a wrapper is ERC-721 token is already widely used and well supported by NFT wallets. And assembly token itself can also be assembled again. Assembly token is easier for smart contract to use than a batch of assets, in scenarios like batch trade, batch swap or collections exchange.
+
+This standard has AssemblyAsset event which records the exact kinds and amounts of assets the assembly token represents. The wallet can easily display those NFTs to user just by the token id.
+
+## Backwards Compatibility
+This proposal combines already available 721 extensions and is backwards compatible with the ERC-721 standard.
+
+## Reference Implementation
```
Comment (lightclient) [inline EIPS/eip-3589.md:83]:
```suggestion
## Implementation
```

File: EIPS/eip-3589.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+## Motivation
+As NFT arts and collectors rapidly increases, some collectors are not satisfied with traditional trading methods. When two collectors want to swap some of their collections, currently they can list their NFTs on the market and notify the other party to buy, but this is inefficient and gas-intensive. Instead, some collectors turn to social media or chat group looking for a trustworthy third party to swap NFTs for them. The third party takes NFTs from both collector A and B, and transfer A's collections to B and B's to A. This is very risky.
+
+The safest way to do batch swap, is to transform batch swap into atomic swap, i.e. one to one swap. But first we should "assemble" those ether, ERC-20 tokens, ERC-721 tokens and ERC-1155 tokens together, and this is the main purpose of this EIP.
+
```
Comment (lightclient) [inline EIPS/eip-3589.md:23]:
```suggestion
```

**Review by lightclient — CHANGES_REQUESTED** [review]

**Comment by lightclient** [comment]
> Need to imporve my english :blush: 

No worries! Welcome :)

**Review by lightclient — APPROVED** [review]

**Pull request merged**